nuxeo-retention/src/main/java/org/nuxeo/retention/service/RetentionManagerImpl.java

```diff
@@ -279,68 +274,92 @@ public void evalRules(Map<String, Set<String>> docsToCheckAndEvents) {
         Framework.getService(WorkManager.class).schedule(work, WorkManager.Scheduling.ENQUEUE);
     }
 
-    protected ELActionContext initActionContext(DocumentModel doc, CoreSession session) {
+    protected boolean evaluateConditionExpression(Record record, String eventInput, CoreSession session) {
+        String expression = record.getRule(session).getStartingPointExpression();
+        if (StringUtils.isEmpty(expression)) {
+            return true;
+        }
         ELActionContext ctx = new ELActionContext(new ExpressionContext(), new ExpressionFactoryImpl());
         ctx.setCurrentPrincipal(session.getPrincipal());
-        doc.detach(true);
-        ctx.setCurrentDocument(doc);
-        return ctx;
-    }
-
-    protected boolean evaluateConditionExpression(ELActionContext ctx, String expression) {
+        record.getDocument().detach(true);
+        ctx.setCurrentDocument(record.getDocument());
         Calendar now = Calendar.getInstance();
-        if (StringUtils.isEmpty(expression)) {
-            return false;
-        }
         ctx.putLocalVariable("currentDate", now);
+        ctx.putLocalVariable(EVENT_BASED_RULE_INPUT, eventInput);
```

Comment:
Need security insight. Do we fear any kind of injection here? Is any escaping needed? FYI, the eventInput is actually a user input...